plugins/spi/SPIOutput.cpp

```diff
@@ -486,6 +502,128 @@ uint8_t SPIOutput::P9813CreateFlag(uint8_t red, uint8_t green, uint8_t blue) {
+    // that can be written as 0xE0 & 0x1F
+    output[spi_offset + 0] = 0xFF;
+    // only write pixel data if buffer has complete data for this pixel:
+    if ((buffer.Size() - offset) >= APA102_SLOTS_PER_PIXEL) {
+      // Convert RGB to APA102 Pixel
+      // skip spi_offset + 0 (is already set)
+      output[spi_offset + 1] = buffer.Get(offset + 2);  // blue
+      output[spi_offset + 2] = buffer.Get(offset + 1);  // green
+      output[spi_offset + 3] = buffer.Get(offset + 0);  // red
+    }
+  }
+  m_backend->Commit(m_output_number);
+}
+
+void SPIOutput::CombinedAPA102Control(const DmxBuffer &buffer) {
+  // for Protocol details see IndividualAPA102Control
+
+  const uint16_t first_slot = m_start_address - 1;  // 0 offset
+
+  // check if enough data is there.
+  if (buffer.Size() - first_slot < APA102_SLOTS_PER_PIXEL) {
+    OLA_INFO << "Insufficient DMX data, required " << APA102_SLOTS_PER_PIXEL
+             << ", got " << buffer.Size() - first_slot;
+    return;
+  }
+
+  // get data for entire string length
+  const uint16_t output_length = (m_pixel_count + 1) *
+                                      APA102_SPI_BYTES_PER_PIXEL;
+  uint8_t *output = m_backend->Checkout(m_output_number, output_length,
+                                      CalculateAPA102LatchBytes(m_pixel_count));
+  // only update SPI data if possible
+  if (!output) {
+    return;
+  }
+
+  // create Pixel Data
+  uint8_t pixel_data[APA102_SPI_BYTES_PER_PIXEL];
+  pixel_data[0] = 0xFF;
+  pixel_data[1] = buffer.Get(first_slot + 2);  // Get Blue
+  pixel_data[2] = buffer.Get(first_slot + 1);  // Get Green
+  pixel_data[3] = buffer.Get(first_slot);  // Get Red
+
+  // set all pixel to same value
+  for (uint16_t i = 0; i < m_pixel_count; i++) {
+    uint16_t spi_offset = APA102_START_FRAME_BYTES +
+                              (i * APA102_SPI_BYTES_PER_PIXEL);
+    memcpy(&output[spi_offset], pixel_data,
+           APA102_SPI_BYTES_PER_PIXEL);
+  }
+
+  // write output back...
+  m_backend->Commit(m_output_number);
+}
+
+/**
+ * Calculate Latch Bytes for APA102:
+ * minimal use half the pixel count bits 
+ * round up to next full byte count.
+ * datasheet says endframe should consist of 4 bytes - 
+ * but thats only valid for up to 64 pixels/leds. (4Byte*8Bit*2=64)
+ *
+ * the function is valid up to 4080 pixels. (255*8*2)
+ * ( otherwise the return type must be changed to uint16_t)
+ */
+uint8_t SPIOutput::CalculateAPA102LatchBytes(uint16_t pixel_count) {
+  // round up so that we get definitely more LatchBits as LEDs/2
+  // const uint8_t latch_bits = ceil(pixel_count / 2);
+  const uint8_t latch_bits = (pixel_count / 2)  + (pixel_count % 2 ? 1: 0);
+  OLA_DEBUG << "latch_bits: " << latch_bits;
+  // same for latch_bytes
+  // const uint8_t latch_bytes = ceil(latch_bits / 8);
+  const uint8_t latch_bytes = (latch_bits / 8) + (latch_bits % 8 ? 1: 0);
+  OLA_DEBUG << "latch_bytes: " << latch_bytes;
+  return latch_bytes;
+}
```

Comment:
Regarding ceil, I'd be tempted to do the cast but I don't know what @nomis52 thinks, although is it not as simple as pixel_count / 2.0? Even with the cast it still feels clearer and easier to read to me; I know exactly what a ceil does when I see it, rather than having to look at your code and go through the various cases to understand that is what it's actually doing. Plus it removes the chance of a future tweak breaking that intention.

I'm not really a fan of temporary variables that are only used once, but that's just my personal preference.